server/utils/hollaex-tools-lib/tools/broker.js

```diff
@@ -24,14 +45,262 @@ const validateBrokerPair = (brokerPair) => {
 	}
 }
 
-const createBrokerPair = (brokerPair) => {
+const getDecimals = (value = 0) => {
+	if (Math.floor(value) === value) return 0;
+
+	let str = value.toString();
+	if (str.indexOf(".") !== -1 && str.indexOf("-") !== -1) {
+		return str.split("-")[1] || 0;
+	} else if (str.indexOf(".") !== -1) {
+		return str.split(".")[1].length || 0;
+	}
+	return str.split("-")[1] || 0;
+}
+
+const binanceScript = async () => {
+	const BINANCE_URL = 'https://api3.binance.com/api/v3/ticker/price'
+
+	// Get the price from redis
+	const formattedSymbol = symbol.split('-').join('').toUpperCase();
+	const quotePrice = await client.getAsync('prices');
+
+	const runScript = (prices) => {
+		//Calculate the price
+		const foundSymbol = JSON.parse(prices).find((data) => data.symbol === formattedSymbol);
+		if (!foundSymbol) {
+			throw new Error('Pair not found');
+		}
+		const baseCurrencyPrice = calculatePrice(foundSymbol.price, side, spread, multiplier);
+
+		const decimalPoint = getDecimals(broker.increment_size);
+		const roundedPrice = math.round(
+			baseCurrencyPrice,
+			decimalPoint
+		);
+
+		const responseObject = {
+			price: roundedPrice
+		}
+		//check if there is user_id, if so, assing token
+		if (user_id) {
+			// Generate randomToken to be used during deal execution
+			const randomToken = generateRandomToken(user_id, symbol, side, broker.quote_expiry_time, roundedPrice);
+			responseObject.token = randomToken;
+		}
+
+		return responseObject;
+	}
+
+	// If it doesn't exist, fetch all market from Binance 
+	if (!quotePrice) {
+		return rp(BINANCE_URL)
+			.then((res) => {
+				//Store all market prices in Redis with 1 minute expiry time
+				//response is a stringfied object.
+				client.setexAsync('prices', 60, res);
+
+				return runScript(res);
+			})
+			.catch(err => {
+				throw new Error(err);
+			})
+
+	} else {
+		return runScript(quotePrice);
+	}
+}
+
+const calculatePrice = (price, side, spread, multiplier = 1) => {
+	// Calculate the price
+	const parsedPrice = parseFloat(price) * multiplier;
+	let calculatedSize;
+
+	if (side === 'buy') {
+		calculatedSize = parsedPrice + (parsedPrice * spread / 100)
+	} else if (side === 'sell') {
+		calculatedSize = parsedPrice - (parsedPrice * spread / 100)
+	}
+
+	return calculatedSize;
+}
+
+const generateRandomToken = (user_id, symbol, side, expiryTime = 30, price) => {
+	// Generate random token
+	const randomToken = randomString({
+		length: 32,
+		numeric: true,
+		letters: true
+	});
+
+	// set the generated token along with trade data in Redis expiry time(quote_expiry_time)
+	const tradeData = {
+		user_id,
+		symbol,
+		price,
+		side,
+	}
+
+	client.setexAsync(randomToken, expiryTime, JSON.stringify(tradeData));
+	return randomToken;
+}
+
+const fetchBrokerQuote = async (brokerQuote) => {
+	const { symbol, side, bearerToken, ip } = brokerQuote;
+
+	try {
+		let user_id = null;
+		if (bearerToken) {
+			const auth = await verifyBearerTokenPromise(bearerToken, ip);
+			if (auth) {
```

Comment:
@fetok12 what happens if the verification fails here?

Reply:
@abeikverdi if it fails, it falls under the catch block. We do not return price and token. it returns this error: "message": "JsonWebTokenError: invalid signature"